File: src/recognizer.cpp
// C++ wrapper for sherpa-onnx offline recognizer
// Uses cpp11 for R interface

#include <sherpa-onnx/c-api/c-api.h>
#include <cpp11.hpp>
#include <memory>
#include <string>
#include <vector>
#include <cstring>
#include <fstream>

using namespace cpp11;

// Deleter function for the recognizer pointer
static void delete_recognizer(const SherpaOnnxOfflineRecognizer *ptr) {
  SherpaOnnxDestroyOfflineRecognizer(ptr);
}

// Validate that a file is a valid WAV file
// Returns true if valid, false otherwise
static bool is_valid_wav(const std::string &filename) {
  std::ifstream file(filename, std::ios::binary);
  if (!file.is_open()) {
    return false;
  }

  // Read RIFF header (first 12 bytes)
  char header[12];
  file.read(header, 12);
  if (!file) {
    return false;
  }

  // Check for "RIFF" magic bytes (0x52494646)
  if (header[0] != 'R' || header[1] != 'I' ||
      header[2] != 'F' || header[3] != 'F') {
    return false;
  }

  // Check for "WAVE" format (0x57415645)
  if (header[8] != 'W' || header[9] != 'A' ||
      header[10] != 'V' || header[11] != 'E') {
    return false;
  }

  return true;
}

// Helper function to convert recognition result to R list
static writable::list convert_result_to_list(const SherpaOnnxOfflineRecognizerResult *result) {
  writable::list out;

  // Text
  out.push_back({"text"_nm = std::string(result->text)});

  // Tokens
  if (result->tokens_arr != nullptr && result->count > 0) {
    writable::strings tokens_vec;
    for (int32_t i = 0; i < result->count; ++i) {
      tokens_vec.push_back(std::string(result->tokens_arr[i]));
    }
    out.push_back({"tokens"_nm = tokens_vec});
  } else {
    out.push_back({"tokens"_nm = R_NilValue});
  }

  // Timestamps
  if (result->timestamps != nullptr && result->count > 0) {
    writable::doubles timestamps_vec;
    for (int32_t i = 0; i < result->count; ++i) {
      timestamps_vec.push_back(result->timestamps[i]);
    }
    out.push_back({"timestamps"_nm = timestamps_vec});
  } else {
    out.push_back({"timestamps"_nm = R_NilValue});
  }

  // Durations
  if (result->durations != nullptr && result->count > 0) {
    writable::doubles durations_vec;
    for (int32_t i = 0; i < result->count; ++i) {
      durations_vec.push_back(result->durations[i]);
    }
    out.push_back({"durations"_nm = durations_vec});
  } else {
    out.push_back({"durations"_nm = R_NilValue});
  }

  // Language
  if (result->lang != nullptr && strlen(result->lang) > 0) {
    out.push_back({"language"_nm = std::string(result->lang)});
  } else {
    out.push_back({"language"_nm = R_NilValue});
  }

  // Emotion
  if (result->emotion != nullptr && strlen(result->emotion) > 0) {
    out.push_back({"emotion"_nm = std::string(result->emotion)});
  } else {
    out.push_back({"emotion"_nm = R_NilValue});
  }

  // Event
  if (result->event != nullptr && strlen(result->event) > 0) {
    out.push_back({"event"_nm = std::string(result->event)});
  } else {
    out.push_back({"event"_nm = R_NilValue});
  }

  // JSON
  if (result->json != nullptr) {
    out.push_back({"json"_nm = std::string(result->json)});
  } else {
    out.push_back({"json"_nm = R_NilValue});
  }

  return out;
}

// Helper function to create a default config
static SherpaOnnxOfflineRecognizerConfig get_default_config() {
  SherpaOnnxOfflineRecognizerConfig config;
  memset(&config, 0, sizeof(config));

  config.feat_config.sample_rate = 16000;
  config.feat_config.feature_dim = 80;
  config.model_config.num_threads = 1;
  config.model_config.debug = 0;
  config.model_config.provider = "cpu";
  config.decoding_method = "greedy_search";
  config.max_active_paths = 4;

  return config;
}

// Create an offline recognizer
// Returns an external pointer to the recognizer
[[cpp11::register]]
SEXP create_offline_recognizer_(
    std::string model_dir,
    std::string model_type,
    std::string encoder_path,
    std::string decoder_path,
    std::string joiner_path,
    std::string model_path,
    std::string tokens_path,
    int num_threads,
    std::string provider,
    std::string language,
    std::string modeling_unit) {

  // Create config
  SherpaOnnxOfflineRecognizerConfig config = get_default_config();

  config.model_config.num_threads = num_threads;
  config.model_config.provider = provider.c_str();
  config.model_config.tokens = tokens_path.c_str();

  // Set modeling_unit if provided (for transducer models)
  if (!modeling_unit.empty()) {
    config.model_config.modeling_unit = modeling_unit.c_str();
  }

  // Set model-specific paths
  if (model_type == "whisper") {
    config.model_config.whisper.encoder = encoder_path.c_str();
    config.model_config.whisper.decoder = decoder_path.c_str();
    config.model_config.whisper.language = language.c_str();
    config.model_config.whisper.task = "transcribe";
    config.model_config.whisper.tail_paddings = -1;
  } else if (model_type == "transducer") {
    config.model_config.transducer.encoder = encoder_path.c_str();
    config.model_config.transducer.decoder = decoder_path.c_str();
    config.model_config.transducer.joiner = joiner_path.c_str();
  } else if (model_type == "paraformer") {
    config.model_config.paraformer.model = model_path.c_str();
  } else if (model_type == "sense-voice") {
    config.model_config.sense_voice.model = model_path.c_str();
    config.model_config.sense_voice.language = language.c_str();
    config.model_config.sense_voice.use_itn = 1;
  } else {
    stop("Unknown model type: %s", model_type.c_str());
  }

  // Create recognizer
  const SherpaOnnxOfflineRecognizer *recognizer =
      SherpaOnnxCreateOfflineRecognizer(&config);

  if (recognizer == nullptr) {
    stop("Failed to create offline recognizer. Please check your model files.");
  }

  // Create external pointer with finalizer
  external_pointer<const SherpaOnnxOfflineRecognizer> ptr(
      recognizer,
      true,
      delete_recognizer);

  return ptr;
}

// Transcribe a WAV file
// Returns a list with transcription results
[[cpp11::register]]
list transcribe_wav_(SEXP recognizer_xptr, std::string wav_path) {
  // Get recognizer from external pointer
  external_pointer<const SherpaOnnxOfflineRecognizer> recognizer(recognizer_xptr);

  if (recognizer.get() == nullptr) {
    stop("Invalid recognizer pointer");
  }

  // Validate WAV file format before processing
  if (!is_valid_wav(wav_path)) {
    stop("Invalid WAV file: %s\nOnly standard WAV files (16-bit PCM, mono/stereo) are supported.\nFile must have RIFF/WAVE headers.", wav_path.c_str());
  }

  // Read WAV file
  const SherpaOnnxWave *wave = SherpaOnnxReadWave(wav_path.c_str());
  if (wave == nullptr) {
    stop("Failed to read WAV file: %s", wav_path.c_str());
  }

  // Create stream
  const SherpaOnnxOfflineStream *stream =
      SherpaOnnxCreateOfflineStream(recognizer.get());

  if (stream == nullptr) {
    SherpaOnnxFreeWave(wave);
    stop("Failed to create offline stream");
  }

  // Accept waveform
  SherpaOnnxAcceptWaveformOffline(
      stream,
      wave->sample_rate,
      wave->samples,
      wave->num_samples);

  // Decode
  SherpaOnnxDecodeOfflineStream(recognizer.get(), stream);

  // Get result
  const SherpaOnnxOfflineRecognizerResult *result =
      SherpaOnnxGetOfflineStreamResult(stream);

  // Convert result to R list using helper
  writable::list out = convert_result_to_list(result);

  // Cleanup
  SherpaOnnxDestroyOfflineRecognizerResult(result);
  SherpaOnnxDestroyOfflineStream(stream);
  SherpaOnnxFreeWave(wave);

  return out;
}

// Transcribe raw audio samples
// Returns a list with transcription results
[[cpp11::register]]
list transcribe_samples_(SEXP recognizer_xptr, doubles samples, int sample_rate) {
  // Get recognizer from external pointer
  external_pointer<const SherpaOnnxOfflineRecognizer> recognizer(recognizer_xptr);

  if (recognizer.get() == nullptr) {
    stop("Invalid recognizer pointer");
  }

  if (samples.size() == 0) {
    stop("Empty audio samples");
  }

  // Convert R doubles to float array
  std::vector<float> samples_vec(samples.size());
  for (size_t i = 0; i < samples.size(); ++i) {
    samples_vec[i] = static_cast<float>(samples[i]);
  }

  // Create stream
  const SherpaOnnxOfflineStream *stream =
      SherpaOnnxCreateOfflineStream(recognizer.get());

  if (stream == nullptr) {
    stop("Failed to create offline stream");
  }

  // Accept waveform
  SherpaOnnxAcceptWaveformOffline(
      stream,
      sample_rate,
      samples_vec.data(),
      samples_vec.size());

  // Decode
  SherpaOnnxDecodeOfflineStream(recognizer.get(), stream);

  // Get result
  const SherpaOnnxOfflineRecognizerResult *result =
      SherpaOnnxGetOfflineStreamResult(stream);

  // Convert result to R list using helper
  writable::list out = convert_result_to_list(result);

  // Cleanup
  SherpaOnnxDestroyOfflineRecognizerResult(result);
  SherpaOnnxDestroyOfflineStream(stream);

  return out;
}

// Destroy a recognizer (explicit cleanup)
[[cpp11::register]]
void destroy_recognizer_(SEXP recognizer_xptr) {
  external_pointer<const SherpaOnnxOfflineRecognizer> recognizer(recognizer_xptr);

  if (recognizer.get() != nullptr) {
    SherpaOnnxDestroyOfflineRecognizer(recognizer.get());
    recognizer.release();
  }
}

// Read a WAV file and return its properties
[[cpp11::register]]
list read_wav_(std::string wav_path) {
  // Validate WAV file format before processing
  if (!is_valid_wav(wav_path)) {
    stop("Invalid WAV file: %s\nOnly standard WAV files (16-bit PCM, mono/stereo) are supported.\nFile must have RIFF/WAVE headers.", wav_path.c_str());
  }

  const SherpaOnnxWave *wave = SherpaOnnxReadWave(wav_path.c_str());

  if (wave == nullptr) {
    stop("Failed to read WAV file: %s", wav_path.c_str());
  }

  // Convert samples to R vector
  writable::doubles samples_vec;
  for (int32_t i = 0; i < wave->num_samples; ++i) {
    samples_vec.push_back(wave->samples[i]);
  }

  writable::list out;
  out.push_back({"samples"_nm = samples_vec});
  out.push_back({"sample_rate"_nm = wave->sample_rate});
  out.push_back({"num_samples"_nm = wave->num_samples});

  SherpaOnnxFreeWave(wave);

  return out;
}

File: R/recognizer.R
# Batch VAD segments together up to a maximum duration
# Each batch will contain concatenated audio from one or more segments
# @param segments List of segments from extract_vad_segments_()
# @param max_duration Maximum batch duration in seconds (default 29 for Whisper)
# @return List of batches, each with samples, start_time, and duration
batch_segments <- function(segments, max_duration = 29.0) {
  if (length(segments) == 0) {
    return(list())
  }

  batches <- list()
  batch_idx <- 1
  seg_idx <- 1

  while (seg_idx <= length(segments)) {
    batch_samples <- numeric(0)
    batch_start_time <- segments[[seg_idx]]$start_time
    batch_duration <- 0.0

    while (seg_idx <= length(segments)) {
      seg <- segments[[seg_idx]]

      # Always add at least one segment; stop if exceeding max
      if (length(batch_samples) > 0 &&
          (batch_duration + seg$duration) > max_duration) {
        break
      }

      batch_samples <- c(batch_samples, seg$samples)
      batch_duration <- batch_duration + seg$duration
      seg_idx <- seg_idx + 1
    }

    batches[[batch_idx]] <- list(
      samples = batch_samples,
      start_time = batch_start_time,
      duration = batch_duration
    )
    batch_idx <- batch_idx + 1
  }

  batches
}

#' Offline Speech Recognizer
#'
#' @description
#' R6 class for offline speech recognition using sherpa-onnx.
#' Supports multiple model architectures including Whisper, Paraformer,
#' SenseVoice, and Transducer models.
#'
#' @importFrom R6 R6Class
#' @importFrom tibble tibble
#' @importFrom utils download.file head
#' @export
OfflineRecognizer <- R6::R6Class(
  "OfflineRecognizer",

  private = list(
    recognizer_ptr = NULL,
    model_info_cache = NULL,
    default_verbose = FALSE,
    num_threads = NULL,
    provider = NULL,

    # Cleanup resources (called automatically on garbage collection)
    finalize = function() {
      if (!is.null(private$recognizer_ptr)) {
        # The external pointer has a finalizer that will call
        # SherpaOnnxDestroyOfflineRecognizer automatically
        private$recognizer_ptr <- NULL
      }
    },

    # Private method for VAD-based transcription
    # Uses vad() for speech detection, then transcribes each batch
    transcribe_with_vad = function(wav_path, vad_config) {
      # Run VAD to detect speech segments
      vad_result <- vad(
        wav_path,
        threshold = vad_config$threshold,
        min_silence = vad_config$min_silence,
        min_speech = vad_config$min_speech,
        max_speech = vad_config$max_speech,
        model = vad_config$model,
        verbose = vad_config$verbose
      )

      # Handle case of no speech detected
      if (vad_result$num_segments == 0) {
        result <- list(
          text = "",
          segments = character(0),
          segment_starts = numeric(0),
          segment_durations = numeric(0),
          num_segments = 0L
        )
        return(new_sherpa_transcription(result, private$model_info_cache))
      }

      # Batch segments (R) - groups segments up to 29s max
      batches <- batch_segments(vad_result$segments, max_duration = 29.0)

      # Transcribe each batch (C++)
      batch_results <- lapply(seq_along(batches), function(i) {
        batch <- batches[[i]]

        if (vad_config$verbose) {
          message(sprintf("Transcribing batch %d: %.2f - %.2f sec",
                          i, batch$start_time, batch$start_time + batch$duration))
        }

        transcription <- transcribe_samples_(
          private$recognizer_ptr,
          batch$samples,
          vad_result$sample_rate
        )

        list(
          text = transcription$text,
          start_time = batch$start_time,
          duration = batch$duration
        )
      })

      # Combine results (R)
      segment_texts <- vapply(batch_results, function(x) x$text, character(1))
      segment_starts <- vapply(batch_results, function(x) x$start_time, numeric(1))
      segment_durations <- vapply(batch_results, function(x) x$duration, numeric(1))

      # Combine text, skipping empty segments
      non_empty <- nzchar(trimws(segment_texts))
      full_text <- trimws(paste(segment_texts[non_empty], collapse = " "))

      result <- list(
        text = full_text,
        segments = segment_texts,
        segment_starts = segment_starts,
        segment_durations = segment_durations,
        num_segments = length(batch_results)
      )

      new_sherpa_transcription(result, private$model_info_cache)
    }
  ),

  public = list(
    #' @description
    #' Create a new offline recognizer
    #'
    #' @param model Model specification. Can be:
    #'   - A shorthand string: "parakeet-v3", "whisper-tiny", "whisper-base", "sense-voice"
    #'   - A HuggingFace repository: "csukuangfj/sherpa-onnx-whisper-tiny.en"
    #'   - A local directory path containing model files
    #' @param language Language code for multilingual models (default: "auto").
    #'   Used for Whisper and SenseVoice models.
    #' @param num_threads Number of threads for inference (default: NULL = auto-detect).
    #'   If NULL, uses parallel::detectCores() with a maximum of 4 threads.
    #' @param provider Execution provider: "cpu", "cuda", or "coreml".
    #'   Default is NULL, which auto-detects: uses "cuda" if available, otherwise "cpu".
    #' @param verbose Logical, whether to print status messages during initialization (default: FALSE).
    #'   This also sets the default verbosity for transcribe() calls.
    #'
    #' @return A new OfflineRecognizer object
    #'
    #' @examples
    #' \dontrun{
    #' # Create recognizer with shorthand (auto-detects threads and provider)
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny")
    #'
    #' # Create recognizer with specific thread count
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny", num_threads = 4)
    #'
    #' # Force CPU even if CUDA is available
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny", provider = "cpu")
    #'
    #' # Create recognizer with HuggingFace repo
    #' rec <- OfflineRecognizer$new(
    #'   model = "csukuangfj/sherpa-onnx-whisper-tiny.en"
    #' )
    #'
    #' # Create recognizer with local model
    #' rec <- OfflineRecognizer$new(model = "/path/to/model")
    #' }
    initialize = function(model = "parakeet-v3",
                         language = "auto",
                         num_threads = NULL,
                         provider = NULL,
                         verbose = FALSE) {

      # Store default verbosity for transcribe() calls
      private$default_verbose <- verbose

      # Auto-detect provider if not specified
      if (is.null(provider)) {
        if (cuda_available()) {
          provider <- "cuda"
          if (verbose) message("CUDA available, using GPU acceleration")
        } else {
          provider <- "cpu"
        }
      }
      private$provider <- provider

      # Auto-detect optimal thread count if not specified
      if (is.null(num_threads)) {
        available_cores <- parallel::detectCores(logical = FALSE)
        # Use up to 4 threads by default (diminishing returns beyond this)
        num_threads <- min(available_cores, 4)
        if (verbose) {
          message(sprintf("Auto-detected %d physical cores, using %d threads",
                         available_cores, num_threads))
        }
      }
      private$num_threads <- num_threads

      # Resolve model
      model_info <- resolve_model(model, verbose = verbose)
      private$model_info_cache <- model_info

      # Get model configuration
      config <- get_model_config(model_info)

      # Prepare paths for C++ function
      encoder_path <- if (!is.null(config$encoder)) config$encoder else ""
      decoder_path <- if (!is.null(config$decoder)) config$decoder else ""
      joiner_path <- if (!is.null(config$joiner)) config$joiner else ""
      model_path <- if (!is.null(config$model)) config$model else ""
      tokens_path <- config$tokens

      # Set modeling_unit for transducer models
      # "cjkchar" is needed for NeMo Parakeet and other CJK models
      modeling_unit <- if (config$model_type == "transducer") "cjkchar" else ""

      # Create recognizer via C++ wrapper
      if (verbose) message("Creating recognizer...")
      private$recognizer_ptr <- create_offline_recognizer_(
        model_dir = config$model_dir,
        model_type = config$model_type,
        encoder_path = encoder_path,
        decoder_path = decoder_path,
        joiner_path = joiner_path,
        model_path = model_path,
        tokens_path = tokens_path,
        num_threads = as.integer(num_threads),
        provider = provider,
        language = language,
        modeling_unit = modeling_unit
      )

      if (verbose) message("Recognizer created successfully")
    },

    #' @description
    #' Transcribe a WAV file
    #'
    #' @param wav_path Path to WAV file (must be 16kHz, 16-bit, mono)
    #' @param verbose Logical. Show progress messages. Default: NULL (inherits from initialize())
    #'
    #' @return A sherpa_transcription object (list-like) containing:
    #'   - text: Transcribed text
    #'   - tokens: Character vector of tokens
    #'   - timestamps: Numeric vector of timestamps (if supported by model)
    #'   - durations: Numeric vector of token durations (if supported by model)
    #'   - language: Detected language (if supported by model)
    #'   - emotion: Detected emotion (if supported by model)
    #'   - event: Detected audio event (if supported by model)
    #'   - json: Full result as JSON string
    #'
    #'   For Whisper models with audio longer than 29 seconds, Voice Activity

    #'   Detection (VAD) is automatically used to segment the audio. In this case,
    #'   additional fields are available:
    #'   - segments: Character vector of segment texts
    #'   - segment_starts: Start times of segments in seconds
    #'   - segment_durations: Duration of segments in seconds
    #'   - num_segments: Number of segments
    #'
    #'   The result has a custom print method but maintains list-like access
    #'   (e.g., `result$text`). Use `as.character(result)` to extract just the
    #'   text, or `summary(result)` for detailed statistics.
    #'
    #' @details
    #' Whisper models have a 30-second context window limit. For audio longer than
    #' 29 seconds, this method automatically uses Voice Activity Detection (VAD)
    #' to split the audio at natural speech boundaries, transcribes each segment,
    #' and combines the results. This happens transparently - you don't need to
    #' configure anything.
    #'
    #' For other model types (Parakeet, SenseVoice, etc.), the entire audio is
    #' transcribed at once regardless of length.
    #'
    #' If you need fine-grained control over VAD parameters, use the standalone
    #' `vad()` function to detect speech segments, then transcribe them individually.
    #'
    #' @examples
    #' \dontrun{
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny")
    #' result <- rec$transcribe("audio.wav")
    #'
    #' # Print with custom format
    #' print(result)
    #'
    #' # Access fields
    #' cat("Transcription:", result$text, "\n")
    #'
    #' # Extract text
    #' text <- as.character(result)
    #'
    #' # Detailed information
    #' summary(result)
    #' }
    transcribe = function(wav_path, verbose = NULL) {
      # Use default verbosity if not specified
      if (is.null(verbose)) {
        verbose <- private$default_verbose
      }

      # Expand tilde and other path shortcuts
      wav_path <- path.expand(wav_path)

      if (!file.exists(wav_path)) {
        stop("WAV file not found: ", wav_path)
      }

      if (is.null(private$recognizer_ptr)) {
        stop("Recognizer not initialized")
      }

      # Check if we need VAD (whisper model + audio > 29s)
      use_vad <- FALSE
      model_type <- private$model_info_cache$model_type

      if (model_type == "whisper") {
        # Read audio to check duration
        wav_data <- read_wav_(wav_path)
        duration <- wav_data$num_samples / wav_data$sample_rate

        if (duration > 29.0) {
          use_vad <- TRUE
          if (verbose) {
            message(sprintf("Audio is %.1f seconds; using VAD for Whisper model", duration))
          }
        }
      }

      # Simple transcription (no VAD needed)
      if (!use_vad) {
        result <- transcribe_wav_(private$recognizer_ptr, wav_path)
        return(new_sherpa_transcription(result, private$model_info_cache))
      }

      # VAD-based transcription with defaults
      vad_config <- list(
        model = "silero-vad",
        threshold = 0.5,
        min_silence = 0.5,
        min_speech = 0.25,
        max_speech = 29.0,
        verbose = verbose
      )

      private$transcribe_with_vad(wav_path, vad_config)
    },

    #' @description
    #' Transcribe multiple WAV files in batch
    #'
    #' @param wav_paths Character vector of WAV file paths
    #'
    #' @return Tibble with one row per file and columns:
    #'   - file: Input file path (character)
    #'   - text: Transcribed text (character)
    #'   - tokens: List-column of token character vectors
    #'   - timestamps: List-column of timestamp numeric vectors (or NULL)
    #'   - durations: List-column of duration numeric vectors (or NULL)
    #'   - language: Detected language (character, NA if not available)
    #'   - emotion: Detected emotion (character, NA if not available)
    #'   - event: Detected audio event (character, NA if not available)
    #'   - json: Full result as JSON string (character)
    #'
    #' @examples
    #' \dontrun{
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny")
    #' results <- rec$transcribe_batch(c("file1.wav", "file2.wav"))
    #'
    #' # Access results via tibble columns
    #' print(results$text)
    #' print(results$file[1])
    #'
    #' # Access list-columns
    #' first_tokens <- results$tokens[[1]]
    #' }
    transcribe_batch = function(wav_paths) {
      if (length(wav_paths) == 0) {
        # Return empty tibble with correct column structure
        return(tibble::tibble(
          file = character(0),
          text = character(0),
          tokens = list(),
          timestamps = list(),
          durations = list(),
          language = character(0),
          emotion = character(0),
          event = character(0),
          json = character(0)
        ))
      }

      # Transcribe all files
      resultstranscribe(wav_paths, function(path) {
        self$transcribe(path)
      })

      # Convert list of results to tibble
      tibble::tibble(
        file = wav_paths,
        text = vapply(results, function(r) r$text, character(1)),
        tokens = lapply(results, function(r) r$tokens),
        timestamps = lapply(results, function(r) r$timestamps),
        durations = lapply(results, function(r) r$durations),
        language = vapply(results, function(r) {
          if (is.null(r$language)) NA_character_ else r$language
        }, character(1)),
        emotion = vapply(results, function(r) {
          if (is.null(r$emotion)) NA_character_ else r$emotion
        }, character(1)),
        event = vapply(results, function(r) {
          if (is.null(r$event)) NA_character_ else r$event
        }, character(1)),
        json = vapply(results, function(r) r$json, character(1))
      )
    },

    #' @description
    #' Get model information
    #'
    #' @return List with model metadata including:
    #'   - type: "local" or "huggingface"
    #'   - path: Local path to model files
    #'   - model_type: Type of model (whisper, paraformer, sense-voice, transducer)
    #'   - repo: HuggingFace repository (if applicable)
    #'
    #' @examples
    #' \dontrun{
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny")
    #' info <- rec$model_info()
    #' print(info)
    #' }
    model_info = function() {
      private$model_info_cache
    },

    #' @description
    #' Print method for OfflineRecognizer
    #'
    #' @param ... Additional arguments (unused)
    #'
    #' @examples
    #' \dontrun{
    #' rec <- OfflineRecognizer$new(model = "whisper-tiny")
    #' print(rec)
    #' }
    print = function(...) {
      info <- private$model_info_cache

      cat("<OfflineRecognizer>\n")

      # Model information
      if (info$type == "huggingface") {
        model_display <- info$repo
        if (!is.null(info$quantization)) {
          model_display <- paste0(model_display, " (", info$quantization, ")")
        }
        cat(sprintf("  Model: %s\n", model_display))
      } else {
        # For local paths, try to extract a meaningful name
        model_name <- basename(info$path)

        # If it looks like a HF cache path (hash), try to get the repo name from parent
        if (grepl("^[a-f0-9]{40}$", model_name)) {
          # Extract from path like models--user--repo/snapshots/hash
          if (grepl("models--", info$path)) {
            parts <- strsplit(info$path, "/")[[1]]
            models_idx <- which(grepl("^models--", parts))
            if (length(models_idx) > 0) {
              # Convert "models--csukuangfj--sherpa-onnx-whisper-tiny.en" to readable format
              repo_part <- parts[models_idx[1]]
              model_name <- gsub("^models--", "", repo_part)
              model_name <- gsub("--", "/", model_name)
            }
          }
        }

        model_display <- paste0(model_name, " (local)")
        if (!is.null(info$quantization)) {
          model_display <- paste0(model_name, " (local, ", info$quantization, ")")
        }
        cat(sprintf("  Model: %s\n", model_display))
      }
      cat(sprintf("  Type: %s\n", info$model_type))

      # Provider and threads
      cat(sprintf("  Provider: %s\n", private$provider))
      cat(sprintf("  Threads: %d\n", private$num_threads))

      # Verbose setting
      cat(sprintf("  Verbose: %s\n", private$default_verbose))

      invisible(self)
    }
  )
)
